### PR TITLE
[RFC] feat: Middlewares addition

### PR DESCRIPTION
**Pull request by petarvujovic98**
Add middlewares decorator
Add middlewares test

Keep in mind that right now the `middleware` decorator needs to be applied after the other function decorators to preserve the capturing of the respective methods. This, however, could be changed if we want to change the way we capture the decorated methods, e.g. by adding a _brand_ property to the decorated functions rather than looking for the name of the function decorator.

Addresses #256

### CODE REVIEW

File: src/near-bindgen.ts [hunk 1]
```diff
@@ -89,6 +89,50 @@ export function call({
   };
 }
 
+/**
+ * The interface that a middleware has to implement in order to be used as a middleware function/class.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+interface Middleware<Arguments extends Array<any>> {
```
Comment (ailisp) [inline src/near-bindgen.ts:96]: Can we make Middleware a `type = Function`, as we don't have other member than `call`. And let user call it as:
```js
@middleware((args) => near.log(args))
```

File: tests/__tests__/test-middlewares.ava.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  const args = { id: "test", accountId: "tset" };
+  // Call the function.
+  const result = await ali.callRaw(middlewares, "get_private", "");
+  // Extract the logs.
+  const { logs } = result.result.receipts_outcome[0].outcome;
+  // Create the expected logs.
+  const expectedLogs = [`Log from middleware: ${args}`];
+
+  t.deepEqual(logs, expectedLogs);
+});
```
Comment (ailisp) [inline tests/__tests__/test-middlewares.ava.js:84]: Great tests that covers compatibility with existing decorators! Could you add a test multiple middleware?

File: src/near-bindgen.ts (hunk 1 above)
Comment (petarvujovic98) [reply src/near-bindgen.ts:96]: @ailisp I can create a commit for this. It's not difficult.

File: tests/__tests__/test-middlewares.ava.js (hunk 2 above)
Comment (petarvujovic98) [reply tests/__tests__/test-middlewares.ava.js:84]: @ailisp Yes, I will add another method where I pass multiple middlewares.

**Comment by petarvujovic98** [comment]
One thing that I am not satisfied with is type inference in the decorator. It would be ideal if I could get the editor/language server to tell the developer what type the middleware arguments will be when the `middleware` decorator is attached to a method. I have not yet seen something similar, but I will try to come up with a solution.

File: src/near-bindgen.ts (hunk 1 above)
Comment (petarvujovic98) [reply src/near-bindgen.ts:96]: @ailisp This is addressed [here](https://github.com/near/near-sdk-js/pull/263/commits/79a9ecbdb3508e2d753dc852ff880c9bcc865c85).

File: tests/__tests__/test-middlewares.ava.js (hunk 2 above)
Comment (petarvujovic98) [reply tests/__tests__/test-middlewares.ava.js:99]: @ailisp This is addressed [here](https://github.com/near/near-sdk-js/pull/263/commits/79a9ecbdb3508e2d753dc852ff880c9bcc865c85).

**Comment by ailisp** [comment]
> One thing that I am not satisfied with is type inference in the decorator. It would be ideal if I could get the editor/language server to tell the developer what type the middleware arguments will be when the middleware decorator is attached to a method. I have not yet seen something similar, but I will try to come up with a solution.

This sounds challenging. I think we can address in future PRs. Your current middleware implementation is general, easy-to-use and well tested. IMO, it's ready for merge.

@volovyks Do you observe any issue regarding Petar's implementation?

**Review by ailisp — APPROVED** [review]

**Review by volovyks — APPROVED** [review]
No objections @ailisp , it is a great implementation.

**Pull request merged**